Add endpoint-specific gnix_fab_req freelist (and new criterion test for endpoints)

sungeunchoi opened the pull request:
I also cleaned up the error exiting in gnix_ep_open().

Note: There is an issue with allocating more than 9 endpoints at a time that needs to be investigated.

@hppritcha @ztiffany @jswaro 

Closes #100 

Signed-off-by: Sung-Eun Choi <sungeunchoi@users.noreply.github.com>

lanl-ompi commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/edison_libfabric_pr/63/
Test PASSed.

jswaro commented:
RE: endpoint allocation

Does it refuse to allocate the endpoint? What is the error code?

sungeunchoi commented:
`GNI_CqCreate` fails with GNI_RC_ERROR_RESOURCE for the rx_cq_blk (which is the 4th call to `GNI_CqCreate` in `gnix_nic_alloc`.

hppritcha commented:
this looks like it needs to be rebased.

hppritcha commented on prov/gni/include/gnix.h:
```diff
@@ -265,8 +266,9 @@ struct gnix_fid_ep {
 	int enabled;
 	int no_want_cqes;
-	/* num. active fab_reqs associated with this ep */
+	/* num. active put (write) fab_reqs associated with this ep */
```
The active count also tracks get (read) reqs.

sungeunchoi commented:
@hppritcha 
I rebased, and also cleaned up the error return path in gnix_ep_open().  FYI, there was a bad check on line 730 (must have missed it during my review of #163).

hppritcha commented:
:100: 

I'm good with this assuming jenkins passes it.

lanl-ompi commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.open-mpi.org/job/edison_libfabric_pr/69/
Test PASSed.
